### ISSUE TITLE
Example application for SSO implementation

### ISSUE THREAD

**Issue by namor** [halo/oauth-sso#1]
Thanks for the detailed document on the implementation of SSO. Very well thought out and pretty much cover all the bases

Do you have an example rails application implementing it?

**Comment by halo** [comment]
Hi! The good news is, this is being used in production and it works. You'd need to look at [this PDF](https://github.com/halo/oauth-sso/raw/master/flow.pdf) though, because that is the latest, most definite specification.

The bad news is that the app is not open source because 90% of it is internal stuff - and not even relevant to SSO. I just didn't have the time to extract the SSO-related code into a separate project.

So where does that leave us? I can answer your questions and provide you with code snippets. Maybe, if you're interested, we could put something together that serves as an example application then. We should maybe start with updating the specification :)

**Comment by namor** [comment]
That would be awesome.. 

Maybe you can extract some of the relevant code pieces based on the steps in the pdf ( there are some missing links )

Eg. Middleware that pulls the grant token & access tokens and writes them on the passport

Then i can piece together a rails doorkeeper app

**Comment by halo** [comment]
Here you go https://gist.github.com/halo/b3f028678eb5e3882d5c

Note that I have 2 warden strategies:

* In the `SessionsController#create` of the SSO server I simply run a `warden.authenticate! :password` to sign in the user on the server in a regular login form.
* In my `Doorkeeper.configure` in the `resource_owner_authenticator` I do the same thing to authenticate immediately if the user does the OAuth dance `warden.user || warden.authenticate!(:password)`.
* In the gist in the `PassportVerification` middleware you'll see the `warden.authenticate! :passport` which just verifies passports sent to us from client apps to verify them.

I took the freedom to give you access to this repository in case you would like to use it as playground and update comments or something. You could even put in an app or engine here and we could work on it together. Sooner or later I need to publish this anyway so why not do it now :)

Oh, just a recommendation, I have the mechanism the client applications use to verify passports in the `lib` directory in the SSO server application along with a gemspec file. That allows me to do full-stack integration tests in one single app. The clients just load the gem and add the warden hook as described in `PassportVerifier`

Don't hesitate to ask questions, I will always answer those asap.

**Comment by halo** [comment]
Oh one more thing I remembered. Since you cannot store a shared oauth secret on an iPhone you would have to make the iphone behave like a browser, i.e. persist the passport in some sort of "cookie" and use that to authenticate every request the iphone makes to a client app. That's the easiest way I think.

**Comment by namor** [comment]
Awesome stuff. That's a lot of code there so i'll have to look through it first to wrap my head around it

The full stack test in a single app is a very nice touch indeed!

**Comment by halo** [comment]
Yes, take your time. As I said, earlier or later I will have to do this anyway, because a new project is awaiting SSO as well. And the code exists and can be extracted whenever you have a missing piece.

One problem I see is that there is no silver bullet since everybody has a different user model, different view logic, etc. It's hard for me to see how to make it reusable. I'm not sure yet we should make an engine or just an example app relying on different modular gems which require minimal coding to get the application running. I think that would be good because we would be somewhat independent Rails versions. I imagine e.g. creating a `doorkeeper-catcher` gem or something to keep it as modular as possible.

Believe me, I had some headaches, too, thinking about all of this :)

**Comment by namor** [comment]
I think the best way to learn this is to implement it step by step... 
Trying to implement Warden.after_authenticate hooks 

Do you have a gist of the relevant part of the Passports model?
Passports.logout
Passports.generate attributes

Its odd though. Isn't it supposed to be named Passport? (following rails model conventions)

**Comment by halo** [comment]
Yes, I will pull them out today. Imagine you simply move all `self.something` methods from a Rails model "Passport" to another module which is called "Passports" just to keep the code cleaner. Whenever you want to interact with a collection of Passports (that is DB querying etc.) you use that plural model.

I even went one step further and do not return Passport instances, but POROs (Plain old ruby objects) so that I effectively force myself not to do something like `user.passports.activations.some_other_association.and_another_one` in a view or something like `Passport.where(some: :thing).limit(4)` in a controller. I just use `Passports.relevant_for_me_just_here` and wrap the logic into that pluralized module. It keeps the business logic clear.
I also only return [operation](https://github.com/halo/operation) instances all the time so that I can predictably know what the results are and it makes testing so much easier.

That's the reason why I cannot simply pull out the code already haha. Too much opinionated conventions! ;) It's my interpretation of the "repository pattern".

**Comment by halo** [comment]
Hm, it's surprisingly simple once translated into normal Rails ;)

```
# app/models/passport.rb
def self.generate(owner_id:, ip:, agent:)
  create owner_id: owner_id, ip: ip, agent: agent, application_id: 0
end

def self.logout(passport_id:, provider_passport_id:)
  client = 0
  count += logout_cluster(passport_id) if passport_id.present?
  count += logout_cluster(provider_passport_id) if provider_passport_id.present?
  count
end

def self.logout_cluster(passport_id)
  return 0 unless passport_id.present? && record = find_by_id(passport_id)
  where(group_id: record.group_id).update_all revoked_at: Time.now, revoke_reason: :logout
end
```

**Comment by halo** [comment]
Actually, here is the entire file, it's not sensitive at all: https://gist.github.com/halo/b3f028678eb5e3882d5c#file-passports-rb

A nice side-effect of using `Operation` is that it works as inline documentation right away :)

**Comment by halo** [comment]
The corresponding model: https://gist.github.com/halo/b3f028678eb5e3882d5c#file-passport-rb

**Comment by halo** [comment]
Do you think we should we go towards a library, an engine or a demo app? I have a hard time deciding.

**Comment by halo** [comment]
So I've been thinking some more about how to have iPhone apps work with SSO. I will try to document my thoughts. It has to do with using OAuth scopes to signify whether the current passport IP address comes directly from the iPhone or whether it can be passed in as a param from a Rails proxy.

Also I think we should go for libraries as much as possible. I.e. small snippets which help you pull out things and make requests and such things. The problem is that every use case is so terribly different. I mean devise for example uses all [these modules](https://github.com/plataformatec/devise/tree/master/lib/devise/models) to be flexible enough for everybody to use it. But I'm not sure we can maintain such flexibility.

Instead I suggest we provide as much reusable code as possible in small gems so you can define your own hooks, callbacks etc. with your use-case specific logic, similar to the way [all these blocks](https://github.com/doorkeeper-gem/doorkeeper/blob/master/lib/generators/doorkeeper/templates/initializer.rb#L7-L13) work in doorkeeper.

We can then have an example rails application which implements all these things to demonstrate how they work together. All the libraries can also be located in the same repository for integration tests.

**Comment by namor** [comment]
Been trying to implement our own implementation on our side.. 
Copy and paste didnt work as some code was missing..

Devise modules make everything configurable which is good but I think devise still implements a good set of conventions which it forces everyone to follow

Maybe you can implement your own conventions as an engine but allow overriding the class to implement user specific logic

as well as before / after_hooks

**Comment by halo** [comment]
Would you mind posting your code somehow? I just grabbed [this beautiful namespace](https://rubygems.org/gems/sso) and I'm starting to implement it, too. As rack middleware, without Rails as far as I can go. That is, when it comes to persistence of the passports, there needs to be some sort of adapter for ActiveRecord I guess.

**Comment by halo** [comment]
Or do you think a Rails engine is a necessity? I think it's only needed for the views, but those are usually overriden anyway.

**Comment by namor** [comment]
Will do tomorrow .. its getting late here :D  Awesome namespace indeed  

I'm not too sure about your rack middleware implementations.   Maybe you can explain your reasoning.
Since you depend on doorkeeper, wouldn't it be better to just hook doorkeeper routes? ..

Eg 
app/middleware/sso/access_token_marker.rb
 def call(env)
      @env = env
      @request = ::ActionDispatch::Request.new @env
      @response = @app.call @env
 
      return response unless request.method == 'POST'
      return response unless authorization_grant_flow? || password_credential_flow?
      return response unless response_code == 200
      return response unless response_body
      return response unless outgoing_access_token


Wouldn't it be better to just hook  the auth route of doorkeeper 
instead of checking for params['code'] ?

**Comment by namor** [comment]
I would use the Rails engine for controller code actually.
If we want to implement our own controllers that would be the way to go  as its easier to implement routes / flow control with engines..

However if we dont need it a gem / libraries would be better actually

**Comment by halo** [comment]
True. I thought that hooking into Doorkeeper internals would be too brittle, but it's true that we depend on it anyway and it depends on Rails in return.

However, I'm not sure Rails is really needed to implement our stuff, except for the persistence. If we only persist passports, I would skip Rails. If we plan on supporting failed logins check, banning, forgot password, failed login attempts limit, etc, then we clearly need Rails.

We should maybe both sleep a night over it ;)

**Comment by namor** [comment]
Sounds like what most rails libraries would do  
```ruby
include sso/railties if Rails
```

Extract out the domain logic into libraries 
then configure the hooks based on Rails / Doorkeeper 

or manually (via Rack or any other method)

**Comment by halo** [comment]
Yeah, I like this.

Hm, I'm not sure how exactly to extract the routes from [in here](https://github.com/doorkeeper-gem/doorkeeper/blob/e24d704fd77ce51d6095d209d4feb195b6ccd8f9/lib/doorkeeper/rails/routes.rb) at runtime, because this thing can be mounted on any subpath. So it might not always be a static `Rails.application.routes.url_helpers.some_path` but e.g. `Rails.application.routes.url_helpers.some_path_somewhere_deep_inside`. Well, maybe it's simpler than that, I just never did it before. 

I'm a little bit concerned about supporting multiple Rails versions, too. I feel a little bit more safe in the middleware stack. And lastly, I thought it might be easier to test multiple rack apps interacting with one another rather than booting a Rails dummy to test against. But again, this problem might only exist in my head :)

**Comment by halo** [comment]
Ok, now it's my time to go to bed :) Time zones are hard.

I'm thinking we might want to create a light-weight [rack-based engine](http://edgeapi.rubyonrails.org/classes/Rails/Engine.html#class-Rails::Engine-label-Endpoint) over [here](https://github.com/halo/sso) and try to get the basics going such as an integration test with a dummy rails app using doorkeeper for pulling out the outgoing tokens.

Good night :)
